### PR TITLE
8307538: Memory leak in TreeTableView when calling refresh

### PR DESCRIPTION
**Pull request by andy-goryachev-oracle**
Fixed a memory leak in TreeTableView by using WeakInvalidationListener (which is the right approach in this particular situation) - the leak is specific to TreeTableRowSkin.

Added a unit test.

Added Refresh buttons and Skin menu to the Monkey Tester (will expand these to other controls as a part next MT update).

### CODE REVIEW

**Comment by andy-goryachev-oracle** [comment]
will separate Monkey Tester changes into a separate PR.

**Comment by hjohn** [comment]
I've seen the problem with Views and Cell creation often enough now that I think we may want to talk about the fundamental problem that is causing these issues.  The lifecycle of a Cell or Row is not the same as the View that contains them. Cells are created on demand, and should either be destroyed or unlinked.

In JavaFX, we have chosen to not have an explicit `destroy` method for cells, but we do have a way to unlink cells.  For a `TreeTableRow` this is a simple as calling `TreeTableRow#updateTreeTableView(null)`. The call already does exactly what you would expect: it unregisters all troublesome listeners (ones that were registered on the view).  The skin also responds to this setting to `null` as it adds listeners to the `treeTableViewProperty` of its skinnable.

In other words, a `TreeTableView` which creates cells on demand with a different lifecycle than itself should also be responsible to clean them up when it no longer needs them -- this is where the real bug is.  It is managing the lifecycle of these cells and should not rely on these cells doing this indirectly by using a lot of weak listeners.  If `TreeTableView` would correctly unlink cells and rows it doesn't need, it would remove the need for **all** weak listeners (and also references) in both `TreeTableRow` and `TreeTableRowSkin`, making their code much simpler, more predictable and easier to test.

Specifically for this PR, only the listeners registered as part of `setupTreeTableViewListeners` need to be weak, the others are not the issue.

**Comment by andy-goryachev-oracle** [comment]
This is a very good suggestion, thanks!

File: modules/javafx.controls/src/test/java/test/javafx/scene/control/skin/TreeTableRowSkinTest.java [hunk 1]
```diff
@@ -208,14 +209,15 @@ public void treeTableRowWithCellSizeShouldHonorPadding() {
...
     @Test
     public void treeTableRowWithFixedCellSizeShouldIgnoreVerticalPadding() {
-        int top = 10;
-        int right = 20;
-        int bottom = 30;
-        int left = 40;
+        int top = 11;
+        int right = 23;
+        int bottom = 37;
+        int left = 49;
```
Comment (kevinrushforth) [inline modules/javafx.controls/src/test/java/test/javafx/scene/control/skin/TreeTableRowSkinTest.java:215]: After the fix and test change, this test now fails for me.

**Review by kevinrushforth — COMMENTED** [review]
If I run just the `TreeTableRowSkinTest` test class, I now get a test failure:

```
$ gradle --info -PTEST_ONLY=true :controls:test --tests TreeTableRowSkinTest
TreeTableRowSkinTest > treeTableRowWithFixedCellSizeShouldIgnoreVerticalPadding() FAILED
    org.opentest4j.AssertionFailedError: expected: <18.0> but was: <23.0>
        at app//org.junit.jupiter.api.AssertionUtils.fail(AssertionUtils.java:55)
        at app//org.junit.jupiter.api.AssertionUtils.failNotEqual(AssertionUtils.java:62)
        at app//org.junit.jupiter.api.AssertEquals.assertEquals(AssertEquals.java:86)
        at app//org.junit.jupiter.api.AssertEquals.assertEquals(AssertEquals.java:81)
        at app//org.junit.jupiter.api.Assertions.assertEquals(Assertions.java:1010)
        at app//test.javafx.scene.control.skin.TreeTableRowSkinTest.treeTableRowWithFixedCellSizeShouldIgnoreVerticalPadding(TreeTableRowSkinTest.java:251)
```

Given that the GHA run passes, as does a full controls test run on my system, that suggests something odd is going on with that test, possibly it is affected by tests in another class. If so, then it could fail randomly anyway, since the tests are run in an unspecified (and unpredictable) order.

**Comment by kevinrushforth** [comment]
@aghaisas can you review this?

**Comment by hjohn** [comment]
> For the purposes of fixing the regression bug, I would like to proceed with this fix as is.

I think that's fine, but I do think you should remove the changes that don't contribute to the fix (I'm pretty sure the weak listeners in the constructor are not a problem).

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/TreeTableRowSkin.java [hunk 2]
```diff
@@ -98,46 +96,45 @@ public TreeTableRowSkin(TreeTableRow<T> control) {
 
         updateTreeItem();
 
-        ListenerHelper lh = ListenerHelper.get(this);
-
-        lh.addChangeListener(control.indexProperty(), (ev) -> {
+        control.indexProperty().addListener(new WeakInvalidationListener((x) -> {
             updateCells = true;
-        });
+        }));
```
Comment (hjohn) [inline modules/javafx.controls/src/main/java/javafx/scene/control/skin/TreeTableRowSkin.java:101]: I don't think this needed changing, the index property is directly available on `TreeTableRow` and so shouldn't cause GC issues.

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/TreeTableRowSkin.java [hunk 3]
```diff
@@ -98,46 +96,45 @@ public TreeTableRowSkin(TreeTableRow<T> control) {
...
 
-        lh.addChangeListener(control.treeItemProperty(), (ev) -> {
+        control.treeItemProperty().addListener(new WeakInvalidationListener((x) -> {
             updateTreeItem();
             // There used to be an isDirty = true statement here, but this was
             // determined to be unnecessary and led to performance issues such as
             // those detailed in JDK-8143266
-        });
+        }));
```
Comment (hjohn) [inline modules/javafx.controls/src/main/java/javafx/scene/control/skin/TreeTableRowSkin.java:108]: As above, I don't think this needs to be weak.

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/TreeTableRowSkin.java [hunk 4]
```diff
@@ -332,7 +331,7 @@ public final DoubleProperty indentProperty() {
 
     private void updateTreeItem() {
         unregisterInvalidationListeners(graphicProperty());
-        treeItem = getSkinnable().getTreeItem();
+        treeItem = (getSkinnable() == null) ? null : getSkinnable().getTreeItem();
```
Comment (hjohn) [inline modules/javafx.controls/src/main/java/javafx/scene/control/skin/TreeTableRowSkin.java:334]: This is not needed if the weak listener change is reverted in the constructor. The cause for it being `null` is that the listener is now no longer removed by `dispose` where before (with `ListenerHelper` or `regsisterXXXListener`) it would have been.

Also, I think when `dispose` is called, this skin should still do its utmost best to remove all listeners immediately.

File: modules/javafx.controls/src/test/java/test/javafx/scene/control/skin/TreeTableRowSkinTest.java (hunk 1 above)
Comment (hjohn) [reply modules/javafx.controls/src/test/java/test/javafx/scene/control/skin/TreeTableRowSkinTest.java:215]: I'm not quite seeing why these changes are needed.  How are they related to the memory leak?

File: modules/javafx.controls/src/test/java/test/javafx/scene/control/skin/TreeTableRowSkinTest.java [hunk 5]
```diff
@@ -320,6 +323,20 @@ public void invisibleColumnsShouldRemoveCorrespondingCellsInRow() {
...
+    /** TreeTableView.refresh() must release all discarded cells JDK-8307538 */
+    @Test
+    public void cellsMustBeCollectableAfterRefresh() {
+        IndexedCell<?> row = VirtualFlowTestUtils.getCell(treeTableView, 0);
+        assertNotNull(row);
+        WeakReference<Object> ref = new WeakReference<>(row);
+        row = null;
+
+        treeTableView.refresh();
+        Toolkit.getToolkit().firePulse();
```
Comment (hjohn) [inline modules/javafx.controls/src/test/java/test/javafx/scene/control/skin/TreeTableRowSkinTest.java:335]: I think cells should also be collectable in other situations, like replacing the row factory.  Is it worth adding a test for that?

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/TreeTableRowSkin.java (hunk 4 above)
Comment (andy-goryachev-oracle) [reply modules/javafx.controls/src/main/java/javafx/scene/control/skin/TreeTableRowSkin.java:334]: The problem is, `dispose()` is never called for cells discarded by the refresh() method.

In the absence of explicit release, these cells are still listening - so using a weak listener in this case is, I think, the right approach.  Adding explicit release per your (absolutely valid) suggestion is a good idea, but much, much more complicated task.

File: modules/javafx.controls/src/test/java/test/javafx/scene/control/skin/TreeTableRowSkinTest.java (hunk 5 above)
Comment (andy-goryachev-oracle) [reply modules/javafx.controls/src/test/java/test/javafx/scene/control/skin/TreeTableRowSkinTest.java:335]: this test reflects the exact failure scenario described in the ticket.

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/TreeTableRowSkin.java (hunk 4 above)
Comment (kevinrushforth) [reply modules/javafx.controls/src/main/java/javafx/scene/control/skin/TreeTableRowSkin.java:334]: Even if you are using weak listeners, `dispose` should remove them.

File: modules/javafx.controls/src/test/java/test/javafx/scene/control/skin/TreeTableRowSkinTest.java (hunk 1 above)
Comment (kevinrushforth) [reply modules/javafx.controls/src/test/java/test/javafx/scene/control/skin/TreeTableRowSkinTest.java:215]: They shouldn't be needed. This points to a problem either with the fix, or a preexisting problem with this test.

File: modules/javafx.controls/src/test/java/test/javafx/scene/control/skin/TreeTableRowSkinTest.java (hunk 5 above)
Comment (kevinrushforth) [reply modules/javafx.controls/src/test/java/test/javafx/scene/control/skin/TreeTableRowSkinTest.java:335]: If it's easy to add another test for additional missing cases that would provide better coverage for your fix, it seems worth doing.

**Review by kevinrushforth — COMMENTED** [review]
I left a couple inline comments.

I also wonder whether the `index` and `treeItem` properties need to be weak (I don't think they do).

The failing test is something that will need to be tracked down. If I revert your changes to `treeTableRowWithFixedCellSizeShouldIgnoreVerticalPadding`, the test passes when that class is run by itself, but fails when the entire controls test suite is run. The modified test fails with that class is run by itself, but passes when the entire controls test suite is run.

This suggests a possible problem where the (now weak) listeners are getting GCed sometimes but not others, and that whether or not it is GCed causes differences that are visible to the test. So either the test is relying on an implementation detail, or there is a functional problem caused by the listeners being GCed too early or not early enough (the latter could point to the need to remove them in dispose).

**Comment by andy-goryachev-oracle** [comment]
it looks like the test is unstable - it works fine in Eclipse, but sometimes fails with gradle.  investigating.

**Comment by kevinrushforth** [comment]
> it looks like the test is unstable - it works fine in Eclipse, but sometimes fails with gradle. investigating.

It's definitely affected by GC, so I'd start by looking there. I can change whether the the test passes or fails by changing the heap size or adding a call to gc before each test.

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/TreeTableRowSkin.java (hunk 4 above)
Comment (andy-goryachev-oracle) [reply modules/javafx.controls/src/main/java/javafx/scene/control/skin/TreeTableRowSkin.java:344]: revered back to using register*Listener, which creates a weak listener and also disconnects them in dispose()

File: modules/javafx.controls/src/test/java/test/javafx/scene/control/skin/TreeTableRowSkinTest.java (hunk 5 above)
Comment (andy-goryachev-oracle) [reply modules/javafx.controls/src/test/java/test/javafx/scene/control/skin/TreeTableRowSkinTest.java:338]: added one more test case.

**Review by hjohn — CHANGES_REQUESTED** [review]
I confirmed the tests fail before the changes in `setupTreeTableViewListeners` and pass with those changes.  Removing the constructor and dispose changes did not fail any tests.

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/TreeTableRowSkin.java [hunk 6]
```diff
@@ -116,43 +109,59 @@ public TreeTableRowSkin(TreeTableRow<T> control) {
...
 
+    private void updateCachedFixedSize() {
+        if (getSkinnable() != null) {
+            TreeTableView<T> t = getSkinnable().getTreeTableView();
+            if (t != null) {
+                DoubleProperty p = t.fixedCellSizeProperty();
+                if (p != null) {
+                    fixedCellSize = p.get();
+                    fixedCellSizeEnabled = fixedCellSize > 0;
+                }
```
[the remark this answers is not in the document]
Comment (andy-goryachev-oracle) [reply modules/javafx.controls/src/main/java/javafx/scene/control/skin/TreeTableRowSkin.java:161]: good point, thanks!

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/TreeTableRowSkin.java [hunk 7]
```diff
@@ -98,13 +93,11 @@ public TreeTableRowSkin(TreeTableRow<T> control) {
...
-        lh.addChangeListener(control.indexProperty(), (ev) -> {
+        registerChangeListener(control.indexProperty(), (x) -> {
             updateCells = true;
         });
 
-        lh.addChangeListener(control.treeItemProperty(), (ev) -> {
+        registerChangeListener(control.treeItemProperty(), (obs) -> {
             updateTreeItem();
             // There used to be an isDirty = true statement here, but this was
             // determined to be unnecessary and led to performance issues such as
```
[the remark this answers is not in the document]
Comment (andy-goryachev-oracle) [reply modules/javafx.controls/src/main/java/javafx/scene/control/skin/TreeTableRowSkin.java:103]: Even though I don't particularly like register*Listener() because of its asymmetric nature (when it comes to removing listeners), here we do need to create weak listeners that get unregistered upon `dispose()`.  We need weak listeners because TreeTableView does not explicitly "disconnect" unused cells (e.g. refresh()), and we need `dispose()` due to skin life cycle.

So, in this particular case, I think this change should be ok.

[the remark this answers is not in the document]
Comment (hjohn) [reply modules/javafx.controls/src/main/java/javafx/scene/control/skin/TreeTableRowSkin.java:103]: Sorry, I think we may be misunderstanding each other.  I'm specifically talking about the two listeners added in `TreeTableRowSkin`'s constructor on `indexProperty` and `treeItemProperty`.  These are part of the `TreeTableRow` which is also discarded when `refresh` is called.

1) The test passes if these changes are reverted. This is a clear indication that either the test is insufficient, or that your assumption, that these must be weak, is incorrect. If you can construct a test that requires these listeners to be weak, then I think the changes are warranted.

2) At the risk of stating the obvious, the listeners in the constructor added using `ListenerHelper` are also correctly removed in `dispose`. `ListenerHelper` does this in a similar way to the `register` methods.

3) The `indexProperty` and `treeItemProperty` listeners are attached to the `TreeTableRow`, not the `TreeTableView`.  The refresh function replaces the entire row, including the skin. There is no need to use weak listeners for this purpose.  Compare this to the listeners that are attached to the `TreeTableView` or the `VirtualFlow`; these have a much longer lifecycle and can survive multiple refreshes and row factory replacements, and hence should be weak (for now).

[the remark this answers is not in the document]
Comment (andy-goryachev-oracle) [reply modules/javafx.controls/src/main/java/javafx/scene/control/skin/TreeTableRowSkin.java:103]: you are right, of course.  and I should know - ListenerHelper is my baby.  I guess the fear of another regression overtook me.

reverting.

thank you for your patience and persistence!

**Review by hjohn — APPROVED** [review]

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/TreeTableRowSkin.java [hunk 8]
```diff
@@ -116,43 +112,57 @@ public TreeTableRowSkin(TreeTableRow<T> control) {
...
             }
         }
     }
 
+    private void updateCachedFixedSize() {
+        if (getSkinnable() != null) {
+            TreeTableView<T> t = getSkinnable().getTreeTableView();
+            if (t != null) {
+                double sz = t.getFixedCellSize();
+                fixedCellSize = sz;
```
Comment (Maran23) [inline modules/javafx.controls/src/main/java/javafx/scene/control/skin/TreeTableRowSkin.java:161]: Why you don't just assign the variable directly?

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/TreeTableRowSkin.java [hunk 9]
```diff
@@ -116,43 +112,57 @@ public TreeTableRowSkin(TreeTableRow<T> control) {
...
+                        }
+                    });
                 }
             }
         }
     }
 
+    private void updateCachedFixedSize() {
+        if (getSkinnable() != null) {
+            TreeTableView<T> t = getSkinnable().getTreeTableView();
```
Comment (Maran23) [inline modules/javafx.controls/src/main/java/javafx/scene/control/skin/TreeTableRowSkin.java:158]: I know this is a short method, but I would rather see a more descriptive variable name here

Comment (andy-goryachev-oracle) [reply modules/javafx.controls/src/main/java/javafx/scene/control/skin/TreeTableRowSkin.java:158]: I prefer not to drag long names if a very descriptive type is right there: `TreeTableView<T> t`

**Review by aghaisas — APPROVED** [review]
The fix looks good to me!

**Review by kevinrushforth — APPROVED** [review]
Looks good.

@hjohn Did you want to re-review (there was only one cleanup change since you approved)?

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/TreeTableRowSkin.java [hunk 10]
```diff
@@ -116,43 +112,56 @@ public TreeTableRowSkin(TreeTableRow<T> control) {
 
     // FIXME: replace listener to fixedCellSize with direct lookup - JDK-8277000
     private void setupTreeTableViewListeners() {
-        ListenerHelper lh = ListenerHelper.get(this);
         TreeTableView<T> treeTableView = getSkinnable().getTreeTableView();
         if (treeTableView == null) {
-            lh.addInvalidationListener(getSkinnable().treeTableViewProperty(), (ev) -> {
+            registerInvalidationListener(getSkinnable().treeTableViewProperty(), (x) -> {
```
Comment (Maran23) [inline modules/javafx.controls/src/main/java/javafx/scene/control/skin/TreeTableRowSkin.java:117]: One question here: Why does this prevent the leak but the ListenerHelper does not?

Comment (andy-goryachev-oracle) [reply modules/javafx.controls/src/main/java/javafx/scene/control/skin/TreeTableRowSkin.java:117]: The difference is that registerInvalidationListener() adds a weak listener, while ListenerHelper adds a strong listener.

It is possible to use ListenerHelper here, at the expense of more complicated code since we'd need to explicitly disconnect the listener when tableViewProperty value gets set.  

Another solution would involve adding a method to add a weak listener to the ListenerHelper to avoid explicit cleanup, or

Go back to the original code which used register/unregister*Listener **in this particular case**.

Comment (Maran23) [reply modules/javafx.controls/src/main/java/javafx/scene/control/skin/TreeTableRowSkin.java:117]: Okay, makes sense.

**Review by Maran23 — APPROVED** [review]
Looks good to me.
